pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -96,90 +100,99 @@ func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, inte
-	for id, repo := range w.gitRepos {
-		branch := repo.GetClonedBranch()
-		w.mu.Lock()
-		err := repo.Pull(ctx, branch)
-		w.mu.Unlock()
-		if err != nil {
-			w.logger.Error("failed to update repository branch",
-				zap.String("repo-id", id),
-				zap.Error(err),
-			)
+// determineUpdates gives back target images to be updated for a given repo.
+func (w *watcher) determineUpdates(ctx context.Context, repoID string, repo git.Repo, provider imageprovider.Provider) ([]config.ImageWatcherTarget, error) {
+	branch := repo.GetClonedBranch()
+	w.mu.Lock()
+	err := repo.Pull(ctx, branch)
+	w.mu.Unlock()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update repository branch: %w", err)
+	}
+
+	// Load Image Watcher Config
+	includes := make([]string, 0)
+	excludes := make([]string, 0)
+	for _, target := range w.config.ImageWatcher.Repos {
+		if target.RepoID != repoID {
 			continue
 		}
+		includes = append(includes, target.Includes...)
+		excludes = append(excludes, target.Excludes...)
+	}
+	cfg, ok, err := config.LoadImageWatcher(repo.GetPath(), includes, excludes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load configuration file for Image Watcher: %w", err)
+	}
+	if !ok {
+		return nil, fmt.Errorf("configuration file for Image Watcher not found: %w", err)
+	}
 
-		includes := make([]string, 0)
-		excludes := make([]string, 0)
-		for _, target := range w.config.ImageWatcher.Repos {
-			if target.RepoID != id {
-				continue
-			}
-			includes = append(includes, target.Includes...)
-			excludes = append(excludes, target.Excludes...)
-		}
-		cfg, ok, err := config.LoadImageWatcher(repo.GetPath(), includes, excludes)
+	updates := make([]config.ImageWatcherTarget, 0)
+	for _, target := range cfg.Targets {
+		outdated, err := checkOutdated(ctx, target, repo, provider)
 		if err != nil {
-			w.logger.Error("failed to load configuration file for Image Watcher", zap.Error(err))
-			continue
+			return nil, fmt.Errorf("failed to check the image is outdated: %w", err)
 		}
-		if !ok {
-			w.logger.Error("configuration file for Image Watcher not found", zap.Error(err))
-			continue
+		if outdated {
+			updates = append(updates, target)
 		}
-		t := filterTargets(provider.Name(), cfg.Targets)
-		targets = append(targets, t...)
 	}
-	return
+	return updates, nil
 }
 
-// filterTargets gives back the targets corresponding to the given provider.
-func filterTargets(provider string, targets []config.ImageWatcherTarget) (filtered []config.ImageWatcherTarget) {
-	for _, t := range targets {
-		if t.Provider == provider {
-			filtered = append(filtered, t)
-		}
+// checkOutdated checks if the image defined in the given target is identical to the one in image provider.
+func checkOutdated(ctx context.Context, target config.ImageWatcherTarget, repo git.Repo, provider imageprovider.Provider) (bool, error) {
+	if provider.Name() != target.Provider {
```

Comment:
nit: could you add notes to clarify why do we treat this case as a not outdated image 🙏

Reply:
As you feel, it's unclear. gonna update them.

Reply:
Updated them.
The Run() spawns runners for each image provider, so the checkOutdated function is called in the context for image provider. We should ignore the targets for other providers.

Reply:
🙆